src/Features/LanguageServer/Protocol/CustomProtocol/FindUsagesLSPContext.cs

```diff
@@ -0,0 +1,220 @@
+                    var classifiedText = definition.GetClassifiedText();
+                    var definitionText = "";
+                    foreach (var text in classifiedText.Runs)
+                    {
+                        definitionText += text.Text;
+                    }
+
+                    // Creating a new VSReferenceItem for the definition
+                    var definitionItem = await GenerateVSReferenceItemAsync(
+                        _id, definitionId: _id, _document, _position, definition.SourceSpans.FirstOrDefault(),
+                        definition.DisplayableProperties, _metadataAsSourceFileService, definitionText,
+                        symbolUsageInfo: null, CancellationToken).ConfigureAwait(false);
+
+                    if (definitionItem.Location != null)
+                    {
+                        AddAndReportResultsIfAtMax(definitionItem);
+                        _id++;
+                    }
+                }
+            }
+        }
+
+        public async override Task OnReferenceFoundAsync(SourceReferenceItem reference)
+        {
+            using (await _semaphore.DisposableWaitAsync(CancellationToken).ConfigureAwait(false))
+            {
+                if (_definitionToId.TryGetValue(reference.Definition, out var definitionId))
+                {
+                    // Creating a new VSReferenceItem for the reference
+                    var referenceItem = await GenerateVSReferenceItemAsync(
+                        _id, definitionId, _document, _position, reference.SourceSpan,
+                        reference.AdditionalProperties, _metadataAsSourceFileService, definitionText: null,
+                        reference.SymbolUsageInfo, CancellationToken).ConfigureAwait(false);
+
+                    if (referenceItem.Location != null)
+                    {
+                        AddAndReportResultsIfAtMax(referenceItem);
+                        _id++;
+                    }
+                }
+            }
+        }
+
+        private void AddAndReportResultsIfAtMax(VSReferenceItem item)
+        {
+            _resultsChunk.Add(item);
+            if (_resultsChunk.Count < MaxResultsChunkSize)
+            {
+                return;
+            }
+
+            _progress.Report(_resultsChunk.ToArray());
+            _resultsChunk.Clear();
+        }
+
+        private static async Task<LSP.VSReferenceItem> GenerateVSReferenceItemAsync(
+            int id,
+            int? definitionId,
+            Document document,
+            int position,
+            DocumentSpan documentSpan,
+            ImmutableDictionary<string, string> properties,
+            IMetadataAsSourceFileService metadataAsSourceFileService,
+            string? definitionText,
+            SymbolUsageInfo? symbolUsageInfo,
+            CancellationToken cancellationToken)
+        {
+            LSP.Location? location = null;
```

Comment:
this is a strong smell that the code isn't clean here.  a great wya to improve things would be:

```c#
var location = await ComputeLocationAsync(...);
if (location == null)
    return null;
```